torchmultimodal/utils/generate.py

```diff
@@ -0,0 +1,212 @@
+            for more details about the required attributes and methods.
+        num_in_tokens (int): Number of input modality tokens.
+        num_out_tokens (int): Number of output modality tokens.
+    """
+
+    def __init__(
+        self,
+        max_seq_len: int,
+        model: nn.Module,
+    ) -> None:
+        if model.training:
+            model = model.eval()
+            warnings.warn(f"{type(model)} is now switched to 'eval' mode.")
+
+        self.max_seq_len = max_seq_len
+        self.model = model
+        self.num_in_tokens = model.num_in_tokens
+        self.num_out_tokens = model.num_out_tokens
+
+    @torch.no_grad()
+    def sample(
+        self,
+        x: Tensor,
+        use_cache: bool = True,
+        causal: bool = False,
+        top_k: Optional[int] = None,
+        top_p: Optional[int] = None,
+        **kwargs: Any,
+    ) -> Tensor:
+        """Generates samples of the output modality based on multinomial distribution.
+
+        Args:
+            x (Tensor): Tensor of batched input, i.e., prompt for the generation.
+            use_cache (bool): If ``True``, key/values of the attention layers will be cached to
+                speed up generation. Defaults to ``True``.
+            causal (bool): If ``True``, use causal attention. Defaults to ``False``.
+            top_k (int, optional): Number of tokens with the highest probability to keep.
+                Defaults to ``None``.
+            top_p (int, optional): Threshold that determines the top tokens to keep in terms of
+                cumulative probability.
+                See py:func:`~torchmultimodal.utils.generate.top_k_top_p_filtering`. Defaults to
+                ``None``.
+
+        Returns:
+            A tensor of generated output modality.
+        """
+        in_tokens = self.model.encode(x, "in", **kwargs)  # type: ignore
+        batch_size, in_seq_len = in_tokens.shape
+        total_seq_len = in_seq_len + self.max_seq_len
+        attn_mask = get_causal_attention_mask(in_seq_len)  # (in_seq_len, in_seq_len)
+        logits_mask = get_logits_mask(
+            in_seq_len, self.max_seq_len, self.num_in_tokens, self.num_out_tokens  # type: ignore
+        )  # (total_seq_len, num_total_tokens)
+
+        # Feed the input modality tokens `(b, in_seq_len)` through the model's transformer to learn
+        # the intermediate context vectors (i.e., key/value).
+        # The sequence is shifted to the right by one unit so that the predicted token at each location
+        # along the sequence is based off the previous token.
+        # Note that the first token is predicted from the learnt start-of-sentence ("sos") token which
+        # gets prepended to the sequence after the position embedding layer.
+        # Attention mask is required to avoid attending to future positions along the sequence.
+        # See :class:`~torchmultimodal.models.gpt.RightShift` for more implementation details.
+        self.model.fwd(  # type: ignore
```

Comment:
No return value?

Reply:
Right this is intentional because for the input modality (or prompt sequence) we are only interested in learning the intermediate key/value representations.

So after calling `fwd` the codepath below will cache the key/value matrices `(batch, head, in_seq_len, d_model)`:

https://github.com/facebookresearch/multimodal/blob/8fc2483290fa03e6b0d8d4ec56a57e43de7406e4/torchmultimodal/modules/layers/attention.py#L218-L231

The cached values will be attended to element-wise during AR decoding for the output-modality.

Might be easier to visualize the internal process by looking at this unit test for GPT model here:

https://github.com/facebookresearch/multimodal/blob/8fc2483290fa03e6b0d8d4ec56a57e43de7406e4/test/models/test_gpt.py#L321-L362

Reply:
Makes sense. Fwd does return something here, right? Maybe just add `_ = self.model.fwd(...)` to make it explicit that we are not using the results.